packages/nodes-base/nodes/Google/Gmail/GenericFunctions.ts

```diff
@@ -0,0 +1,126 @@
+import {
+	OptionsWithUri,
+} from 'request';
+
+import {
+	IExecuteFunctions,
+	IExecuteSingleFunctions,
+	ILoadOptionsFunctions,
+} from 'n8n-core';
+
+import {
+	IDataObject,
+} from 'n8n-workflow';
+
+export async function googleApiRequest(this: IExecuteFunctions | IExecuteSingleFunctions | ILoadOptionsFunctions,
+	method: string, resource: string, body: any = {}, qs: IDataObject = {}, uri?: string,
+	option: IDataObject = {}): Promise<any> { // tslint:disable-line:no-any
+	let options: OptionsWithUri = {
+		headers: {
+			'Accept': 'application/json',
+			'Content-Type': 'application/json',
+		},
+		method,
+		body,
+		qs,
+		uri: uri || `https://www.googleapis.com${resource}`,
+		json: true,
+	};
+
+	options = Object.assign({}, options, option);
+
+	try {
+		if (Object.keys(body).length === 0) {
+			delete options.body;
+		}
+
+		//console.log(options);
+
+		//@ts-ignore
+		return await this.helpers.requestOAuth2.call(this, 'gmailOAuth2', options);
+
+	} catch (error) {
+		if (error.response && error.response.body && error.response.body.error) {
+
+			let errorMessages;
+
+			if (error.response.body.error.errors) {
+				// Try to return the error prettier
+				errorMessages = error.response.body.error.errors;
+
+				errorMessages = errorMessages.map((errorItem: IDataObject) => errorItem.message);
+
+				errorMessages = errorMessages.join('|');
+
+			} else if (error.response.body.error.message) {
+				errorMessages = error.response.body.error.message;
+			}
+
+			throw new Error(`Gmail error response [${error.statusCode}]: ${errorMessages}`);
+		}
+		throw error;
+	}
+}
+
+export function encodeEmail(email: any = {}) { // I feel like I should not have to use any here
```

Comment:
After the colon you should describe the shape of the object that you'll pass into `encodeEmail()`.

This is just an arbitrary example without looking at your code in detail:

```ts
type myEmail = {
	to: string; // a property called `to` of type `string` _must_ be present
	from: string;
	mimeVersion: number;
	cc?: Array<string>; // the property `cc` of type `Array<string>` may or may not be present
}
```
And then:

```ts
export function encodeEmail(email: myEmail) {
    // ...
}
```

Or you can insert the whole description after the colon, without naming the type. Interfaces are used similarly.

This way, if you try to pass in an object without the required properties, or with the required properties but with wrong types, TypeScript will let you know before you even run your code.

Reply:
@ivov Thanks for the initial feedback!! I just renamed according to conventions, got rid of the <> around message-id for easier use, and replaced `0` with `i` in a few places. I also created interfaces to get rid of the TS errors.